text/0000-pattern-3.md

```diff
@@ -0,0 +1,1653 @@
+                ^___________________^                ^                   ^
+                 3..8                                3                   8
+
+split_around:           ^_______^                            ^_______^
+                         5..7                                 2..4
+
+               -+---+---+---+---+---+-               +---+---+
+                | C | D | E | F | G |                | C | D |
+               -+---+---+---+---+---+-               +---+---+
+                ^_______^                            ^       ^
+                 3..5                                3       5
+
+               -+---+---+---+---+---+-                       +---+---+
+                | C | D | E | F | G |                        | E | F |
+               -+---+---+---+---+---+-                       +---+---+
+                        ^_______^                            ^       ^
+                         5..7                                5       7
+
+               -+---+---+---+---+---+-                               +---+
+                | C | D | E | F | G |                                | G |
+               -+---+---+---+---+---+-                               +---+
+                                ^___^                                ^   ^
+                                 7..8                                7   8
+```
+
+A span of shared haystack will always store a copy of the original haystack when splitting,
+because the haystack can be cheaply cloned. Splitting is thus just manipulation of the range only.
+Slicing is only done when returning from an algorithm.
+
+A non-shared haystack needs to maintain unique ownership for each haystack slice. Therefore,
+a haystack will be split as soon as the span is split. The "original range" becomes a value
+disconnected from the haystack, and this is where `.restore_range()` is needed:
+to recover the indices in the middle (`5 == 3 + 2` and `7 == 3 + 4`).
+
+### Searcher
+
+A searcher has two required methods `.search()` and `.consume()`,
+and an optional method `.trim_start()`.
+
+```rust
+pub unsafe trait Searcher<A: Hay + ?Sized> {
+    fn search(&mut self, span: Span<&A>) -> Option<Range<A::Index>>;
+    fn consume(&mut self, span: Span<&A>) -> Option<A::Index>;
+    fn trim_start(&mut self, hay: &A) -> A::Index { ... }
+}
+
+pub unsafe trait ReverseSearcher<A: Hay + ?Sized>: Searcher<A> {
+    fn rsearch(&mut self, span: Span<&A>) -> Option<Range<A::Index>>;
+    fn rconsume(&mut self, span: Span<&A>) -> Option<A::Index>;
+    fn trim_end(&mut self, hay: &A) -> A::Index { ... }
+}
+
+pub unsafe trait DoubleEndedSearcher<A: Hay + ?Sized>: ReverseSearcher<A> {}
+```
+
+`.search()` and `.consume()` are safe because there is no safe ways to construct a `Span<&A>`
+with invalid ranges. Implementations of these methods often start with:
+
+```rust
+    fn search(&mut self, span: SharedSpan<&A>) -> Option<Range<A::Index>> {
```

Comment:
Is `SharedSpan` a relic of a previous version of this proposal? I don't see it defined anywhere and it sounds like `Span<H> where H: SharedHaystack`.

Reply:
Fixed. (In an ancient version there was `SharedSpan<H>` and `UniqueSpan<H>` where `Haystack` has an associated type `Span` to determine which span to use. The resulting code was quite ugly.)